src/components/Navbar/index.spec.tsx

```diff
@@ -191,4 +212,213 @@ describe(__filename, () => {
       expect(root.find(CommentSummaryButton)).toHaveLength(1);
     });
   });
+
+  describe('loading data', () => {
+    const renderForFetching = ({
+      store = configureStore(),
+      ...moreProps
+    }: RenderParams = {}) => {
+      const dispatch = spyOn(store, 'dispatch');
+      const fakeThunk = createFakeThunk();
+      const _fetchVersion = fakeThunk.createThunk;
+
+      const props = { _fetchVersion, store, ...moreProps };
+      const root = render(props);
+
+      return { _fetchVersion, dispatch, root, fakeThunk };
+    };
+
+    it('loads data on mount', () => {
+      const loadData = jest.spyOn(NavbarBase.prototype, 'loadData');
+
+      renderForFetching();
+
+      expect(loadData).toHaveBeenCalled();
+    });
+
+    it('loads data on update', () => {
+      const loadData = jest.spyOn(NavbarBase.prototype, 'loadData');
+
+      const { root } = renderForFetching();
+      loadData.mockClear();
+
+      root.setProps({});
+
+      expect(loadData).toHaveBeenCalled();
+    });
+
+    it('fetches the base version when it has an ID and a current version', () => {
+      const addonId = nextUniqueId();
+      const store = createStoreWithCurrentVersion({ addonId });
+
+      const currentBaseVersionId = nextUniqueId();
+      store.dispatch(
+        versionsActions.setCurrentBaseVersionId({
+          versionId: currentBaseVersionId,
+        }),
+      );
+
+      const { _fetchVersion, dispatch, fakeThunk } = renderForFetching({
+        store,
+      });
+
+      expect(dispatch).toHaveBeenCalledWith(fakeThunk.thunk);
```

Comment:
I find this a bit confusing because `fakeThunk` is not defined anywhere in this test, but rather in the `renderForFetching` helper. This means I have no idea what we're verifying here, without having to look it up in `renderForFetching`. I agree that the helper is quite useful, but I feel like this goes a bit against our approach of trying to set up data that will be asserted upon in the test case itself. I'm not sure if there is a better solution to this problem, but figured I'd raise it.

Reply:
I see what you mean. Yeah, we typically make sure that anything used in an assertion is passed as a direct input to the helper.

The first option for fixing it is to make all inputs explicit, like:

<details>
<summary>Patch</summary>

```diff
diff --git a/src/components/Navbar/index.spec.tsx b/src/components/Navbar/index.spec.tsx
index 0228e6c..b44bf22 100644
--- a/src/components/Navbar/index.spec.tsx
+++ b/src/components/Navbar/index.spec.tsx
@@ -215,12 +215,15 @@ describe(__filename, () => {
 
   describe('loading data', () => {
     const renderForFetching = ({
+      fakeThunk = createFakeThunk(),
+      _fetchVersion = fakeThunk.createThunk,
       store = configureStore(),
       ...moreProps
-    }: RenderParams = {}) => {
+    }: {
+      fakeThunk?: ReturnType<typeof createFakeThunk>;
+      _fetchVersion?: ReturnType<typeof createFakeThunk>['createThunk'];
+    } & RenderParams = {}) => {
       const dispatch = spyOn(store, 'dispatch');
-      const fakeThunk = createFakeThunk();
-      const _fetchVersion = fakeThunk.createThunk;
 
       const props = { _fetchVersion, store, ...moreProps };
       const root = render(props);
@@ -258,7 +261,12 @@ describe(__filename, () => {
         }),
       );
 
-      const { _fetchVersion, dispatch, fakeThunk } = renderForFetching({
+      const fakeThunk = createFakeThunk();
+      const _fetchVersion = fakeThunk.createThunk;
+
+      const { dispatch } = renderForFetching({
+        _fetchVersion,
+        fakeThunk,
         store,
       });
 
```

</details>

However, this means that any assertion using `fakeThunk` also has to define `_fetchVersion` even if they don't use it in the assertion. That seems a little excessive because when `_fetchVersion` isn't used in an assertion, the reader has to ponder it unnecessarily.

I decided to return `_fetchVersionThunk` for use in assertions since that provides a much better hint to the reader about what it *is* which should alleviate the original problem you raised of having to look up the definition to figure it out.

<details>
<summary>Patch</summary>

```diff
diff --git a/src/components/Navbar/index.spec.tsx b/src/components/Navbar/index.spec.tsx
index 0228e6c..ea67a88 100644
--- a/src/components/Navbar/index.spec.tsx
+++ b/src/components/Navbar/index.spec.tsx
@@ -225,7 +225,12 @@ describe(__filename, () => {
       const props = { _fetchVersion, store, ...moreProps };
       const root = render(props);
 
-      return { _fetchVersion, dispatch, root, fakeThunk };
+      return {
+        _fetchVersion,
+        _fetchVersionThunk: fakeThunk.thunk,
+        dispatch,
+        root,
+      };
     };
 
     it('loads data on mount', () => {
@@ -258,11 +263,13 @@ describe(__filename, () => {
         }),
       );
 
-      const { _fetchVersion, dispatch, fakeThunk } = renderForFetching({
-        store,
-      });
+      const { _fetchVersion, _fetchVersionThunk, dispatch } = renderForFetching(
+        {
+          store,
+        },
+      );
 
-      expect(dispatch).toHaveBeenCalledWith(fakeThunk.thunk);
+      expect(dispatch).toHaveBeenCalledWith(_fetchVersionThunk);
       expect(_fetchVersion).toHaveBeenCalledWith({
         addonId,
         path: undefined,
@@ -273,16 +280,16 @@ describe(__filename, () => {
 
     it('does not fetch the base version without a current version', () => {
       const store = createStoreWithVersion({ makeCurrent: false });
-      const { dispatch, fakeThunk } = renderForFetching({ store });
+      const { _fetchVersionThunk, dispatch } = renderForFetching({ store });
 
-      expect(dispatch).not.toHaveBeenCalledWith(fakeThunk.thunk);
+      expect(dispatch).not.toHaveBeenCalledWith(_fetchVersionThunk);
     });
 
     it('does not fetch the base version without an ID', () => {
       const store = createStoreWithVersion({ makeCurrent: true });
-      const { dispatch, fakeThunk } = renderForFetching({ store });
+      const { _fetchVersionThunk, dispatch } = renderForFetching({ store });
 
-      expect(dispatch).not.toHaveBeenCalledWith(fakeThunk.thunk);
+      expect(dispatch).not.toHaveBeenCalledWith(_fetchVersionThunk);
     });
 
     it('does not fetch the base version when it already exists', () => {
@@ -299,9 +306,9 @@ describe(__filename, () => {
           versionId: currentBaseVersionId,
         }),
       );
-      const { dispatch, fakeThunk } = renderForFetching({ store });
+      const { _fetchVersionThunk, dispatch } = renderForFetching({ store });
 
-      expect(dispatch).not.toHaveBeenCalledWith(fakeThunk.thunk);
+      expect(dispatch).not.toHaveBeenCalledWith(_fetchVersionThunk);
     });
 
     it('does not fetch the base version when there was an error fetching it', () => {
@@ -317,9 +324,9 @@ describe(__filename, () => {
       store.dispatch(
         versionsActions.abortFetchVersion({ versionId: currentBaseVersionId }),
       );
-      const { dispatch, fakeThunk } = renderForFetching({ store });
+      const { _fetchVersionThunk, dispatch } = renderForFetching({ store });
 
-      expect(dispatch).not.toHaveBeenCalledWith(fakeThunk.thunk);
+      expect(dispatch).not.toHaveBeenCalledWith(_fetchVersionThunk);
     });
   });

 
```

</details>

Thanks for catching that. Let me know if you have further thoughts on improving it.

Reply:
I think that's a good approach, thanks!